torch/onnx/symbolic_opset11.py

```diff
@@ -143,9 +143,14 @@ def index_put(g, self, indices_list_value, values, accumulate=False):
         dtype = sym_help.scalar_type_to_onnx.index(sym_help.cast_pytorch_to_onnx[dtype])
         dtype = sym_help.scalar_type_to_pytorch_type[dtype]
         zeros = g.op("ConstantOfShape", g.op("Shape", self), value_t=torch.tensor([0], dtype=dtype))
```

Comment:
it appears you will need to cast self to dtype of values for both cases, if they don't match.

For zeros, it is constructed from ConstantOfShape. So instead of casting later, you can construct it with the decided dtype.